### PR TITLE
Unify storage base path and fix background image URLs

### PR DESCRIPTION
**Pull request by DasSkelett**
Sorry, this ended up way bigger than I hoped, due to the background image URL issue.

## Problems
Our content path handling is currently all over the place. Every function that saves or accesses some user content from disk has its own implementation to find the location.

Profile background images are not saved in the user folder, but instead we create a new folder per upload, with the timestamp in the folder name. When a profile background is deleted, the folder stays behind, empty:
```
├── HebaruSan-1622467520.5473514_1
├── HebaruSan-1622467781.61281_1
├── HebaruSan-1622467793.0606115_1
│   └── HebaruSan.jpg
├── HebaruSan_1
│   └── CKAN_Publish_Test
│       └── CKAN_Publish_Test-12345.zip
```

---

The background image URL handling is pretty awful right now:
When uploading background images (mods and profiles), we first save the storage-relative path in the db (like we do for mod zips).
https://github.com/KSP-SpaceDock/SpaceDock/blob/ebb54cb85da0ea798cdbc3578e52b56fd04ee0c2/KerbalStuff/blueprints/api.py#L483
But to the user, we return a path prefixed with a hard-coded `/content/`.
https://github.com/KSP-SpaceDock/SpaceDock/blob/ebb54cb85da0ea798cdbc3578e52b56fd04ee0c2/KerbalStuff/blueprints/api.py#L484
Then, when the user hits "Save Changes", we overwrite the path in the db with this `/content/...` path.
https://github.com/KSP-SpaceDock/SpaceDock/blob/ebb54cb85da0ea798cdbc3578e52b56fd04ee0c2/KerbalStuff/blueprints/profile.py#L83
And in the template, we embed this `/content/...` URL as is, which makes everything work again.
https://github.com/KSP-SpaceDock/SpaceDock/blob/ebb54cb85da0ea798cdbc3578e52b56fd04ee0c2/templates/profile.html#L11

The `/content/` prefix is not something the backend should ever deal with. We should be able to set whatever `cdn-domain` we want in `config.ini` (e.g. `spacedock.bigcdnprovider.com/usercontent/`), without the site breaking because it tries to point everyone to `/content/`.

---

When deleting a mod, it doesn't delete the referral events. Also we delete the mod first, before deleting rows of other tables referencing it, which theoretically could lead to those other rows not being removed because the foreign key was updated to NULL beforehand. Fortunately, SqlAlchemy seems to reorder the statements so they work:
```sql
UPDATE referralevent SET mod_id=NULL WHERE referralevent.id = 18
DELETE FROM featured WHERE featured.id = 4
UPDATE mod SET default_version_id=NULL WHERE mod.id = 10014
DELETE FROM modversion WHERE modversion.id = 100021
DELETE FROM modversion WHERE modversion.id = 100022
DELETE FROM mod WHERE mod.id = 10014
COMMIT
```

## Changes

There is a  new `User.base_path()` and a new `Mod.base_path()` method, which returns the base path for all user content relative to the storage root (e.g. `user_1` `user_2/mod_name`).
They are used wherever we need to store or access files in the storage.

Profile images are now saved in the user folder (`User.base_path()`), the format of the file names is `{username}-header-{upload_time}`, where `upload_time` no longer is a float to the nanoseconds, but an int to the seconds.

---

When clicking "Save Changes" after updating a background image for your profile or a mod, the background image path is no longer overwritten, but stays in the storage-relative form. 
Old data in the db with `/content/` prefixed is fixed in the migration.

This means that we need to translate the image paths to proper URLs before embedding the links into the HTML.
This is done in new `Mod.background_url()` and `User.background_url()` methods. They either return the URL concatenated with `cdn-domain` if it's set, or fall back to new `/mods/<mod_id>/<mod_name>/background` or `/profile/<username>/background` routes. That's slightly different to how we handle it for mod downloads, where we always point to the `/mod/<mod_id>/<mod_name>/download` URL, and then redirect based on settings. But this redirection would break client-side caching, which I definitely want to keep, especially for thumbnails.
`thumbnail.get_or_create()`, which is used to calculate the thumbnail URLs from the background image paths is adjusted accordingly, and also returns an URL either for the CDN or to the new `/mods/<mod_id>/<mod_name>/thumb` route.
We also save the thumbnail path in the database now, to remove the `isfile` call to the storage.

These three new routes can serve the files via X-Sendfile, if configured. The "X-Sendfile with fallback to serve-from-gunicorn" logic has been factored out into a common `sendfile()` function.
Since it also sees images now, the `Content-Type` header is no longer hard-coded, but instead we use the Python stdlib `mimetypes` package to guess the mime type _from the file extension_. We definitely want to avoid a disk access there, even if it means the file type could theoretically be incorrect (it's good enough for our use case).

To simplify testing all this, I've also committed a change to `nginx.nginx`, so it can handle X-Sendfile.

---

ReferralEvents that reference a mod that's about to be deleted are removed from the database as well.
The mod deletion now happens after all the other deletes, so we don't rely on SqlAlchemy being smart:
```sql
DELETE FROM featured WHERE featured.id = 5
DELETE FROM referralevent WHERE referralevent.id = 19
UPDATE mod SET default_version_id=NULL WHERE mod.id = 10015
DELETE FROM modversion WHERE modversion.id = 100023
DELETE FROM modversion WHERE modversion.id = 100024
DELETE FROM mod WHERE mod.id = 10015
COMMIT
```

The API/JSON error handling is updated to no longer leak potentially confidential error messages, if the server is not running in debug mode.

Aaand the PURGE request to purge cached zips from ATS when a mod release is deleted is now surrounded by try-catch, since it fails in the docker dev environment with nginx, and it should also not cancel the deletion if Apache complains on production.

## Deployment note
Run the following command after this is deployed on production, to get rid of all existing empty profile background directories:
```bash
find /storage/sdmods/ -empty -type d -delete
```

### CODE REVIEW

**Comment by HebaruSan** [comment]
> This could be solved in a migration, but I didn't want to add one here.

If I'm understanding the changes correctly, it would be at most one changed row per mod and one per user, right? Probably mostly mods, as few users set backgrounds (and especially not our spammy advertising friends that fill most of the db)?

File: KerbalStuff/blueprints/api.py
Comment (HebaruSan) [inline KerbalStuff/blueprints/api.py:187]:
```suggestion
    base_path = os.path.join(current_user.base_path(), mod_name_sec)
```

Do we need the `f'{}'` here? Looks like a single value returned with no extra formatting.

**Comment by DasSkelett** [comment]
> If I'm understanding the changes correctly, it would be at most one changed row per mod and one per user, right? 

Yup. Though we still need to check each row to determine whether it needs updating or not.

File: KerbalStuff/blueprints/api.py
Comment (DasSkelett) [reply KerbalStuff/blueprints/api.py:187]: Good spot, a left-over from before.

File: KerbalStuff/blueprints/api.py
Comment (HebaruSan) [inline KerbalStuff/blueprints/api.py:480]: None of the f-string tutorials I found explained what the `!s` does. I had to read the official PEP document.

Comment (DasSkelett) [reply KerbalStuff/blueprints/api.py:480]: Same, I could also only find information on it in the PEP. But as far as I remember it said it was deprecated, so I didn't keep it. I'm not sure if I've ever used these flags anywhere, the implicit string conversion usually does what I want.

**Comment by HebaruSan** [comment]
> Yup. Though we still need to check each row to determine whether it needs updating or not.

It's a very delicate dance each upgrade, trying to fill VITAS's tolerance level for migration delays up to the maximum without overflowing. :scientist::straight_ruler::tumbler_glass:

File: KerbalStuff/thumbnail.py
Comment (HebaruSan) [inline KerbalStuff/thumbnail.py:56]:
```suggestion
def get_or_create(mod: 'Mod'):
```

What about a forward reference? Would that work?

File: KerbalStuff/objects.py
Comment (HebaruSan) [inline KerbalStuff/objects.py:88]: In 4 places we have `[9:]`, in 2 places we have `[8:]`. Is the inconsistency intentional?

**Comment by HebaruSan** [comment]
> When a profile background is deleted, the folder stays behind, empty

Should we think about cleaning those up? I suppose there may not be enough of them to matter, since most of the time a user who picks a background would keep the same one for a long time.

File: KerbalStuff/blueprints/api.py
Comment (HebaruSan) [reply KerbalStuff/blueprints/api.py:480]: Yeah, it looks like it's a no-op.

File: KerbalStuff/objects.py
Comment (DasSkelett) [reply KerbalStuff/objects.py]: Should be the same, of course. I could've sworn I changed those two to 9 at some point...

File: KerbalStuff/thumbnail.py
Comment (DasSkelett) [reply KerbalStuff/thumbnail.py]: The answer was a mix of this string annotation and a conditional import behind the global `TYPE_CHECKING` variable:
https://adamj.eu/tech/2021/05/13/python-type-hints-how-to-fix-circular-imports/

**Comment by DasSkelett** [comment]
> Should we think about cleaning those up? I suppose there may not be enough of them to matter, since most of the time a user who picks a background would keep the same one for a long time.

Removing the folders of images that get deleted/replaced after this PR is deployed is easy, I'm going to add that.
But I don't know where code for removing existing empty directories should be put. It would be a one-time action. Maybe as command in the `./spacedock` script?

**Comment by HebaruSan** [comment]
> But I don't know where code for removing existing empty directories should be put.

I don't know, either. It's a migration outside of the database. Would it be excusable to have alembic handle it?

**Comment by DasSkelett** [comment]
> Would it be excusable to have alembic handle it?

I'm not too fond of doing IO on the storage mount during database upgrades, if the mount goes wonky who knows how the database migration behaves.
Additionally this is something we could do at any time, it doesn't need to be while everything's down, we could save us a few seconds or minutes of downtime there.

Maybe we need to come up with a new non-db non-downtime storage "migration" system?

**Comment by HebaruSan** [comment]
Yeah, that makes sense. Maybe just have a `find` command prepared for @V1TA5 to run?

**Comment by DasSkelett** [comment]
`find /storage/sdmods/  -maxdepth 1 -empty -type d -delete`
removes all top-level empty directories, which should be mostly empty profile image dirs, and maybe some user dirs if a user had mods in the past but deleted them all.

`find /storage/sdmods/ -empty -type d -delete` 
removes all empty directories recursively, probably doesn't hurt to run either. Can't think of any empty directory we'd want to keep.

I'd take the second one, if you don't prefer otherwise. I'm gonna put it into the PR description, and maybe also apply the `Type: Migration` label so we don't forget about it?

**Comment by HebaruSan** [comment]
Checking some recent profiling, there are `/kerbal-space-program/browse/updated` and `/kerbal-space-program` runs that slow down in a `render_template` → ... → `background_thumb` → `get_or_create` → `isfile` call stack. So every time we render a "mod box" in a template, we are checking whether the thumbnail exists before returning the HTML. Can we get rid of that and just send whatever is stored in SQL as the thumbnail without checking storage?

**Comment by DasSkelett** [comment]
> Can we get rid of that and just send whatever is stored in SQL as the thumbnail without checking storage?

The thumbnail path isn't stored in the database like the background paths, instead it's constructed on the fly by prefixing a `thumb_` to the filename, and for this it needs to check whether the file already exists or whether it needs to generate it.

I'm basically seeing two ways to solve that:
1) Generate the thumbnail instantly when uploading the background image. In this case we can still just prefix a `thumb_`, but assume that it exists without checking it.
2) Create a new column for the thumbnail path, and save the path in the db when generating it, then return the path without checking for the file. Could also be combined with 2, theoretically.

**Comment by HebaruSan** [comment]
Yeah, let's create the thumbnail at upload. I don't think I care much whether we store it in a new SQL column or not, so I'd leave that up to you.

**Comment by DasSkelett** [comment]
Okay. That leaves only the question how we're dealing with potential mods that are uploaded shortly before deploying this code, and don't have a thumbnail generated yet.
The possibility is small, as long as someone looks at `/kerbal-space-program`, `/kerbal-space-program/browse` or `/kerbal-space-program/browse/new` in the meantime they'll be generated.

Should we just ignore it and hope for the best? Or write yet another manual migration script to run afterwards? That one will be very complex though, since we need to access the database to get the newest mods, and then call the thumbnail generation code.

**Comment by HebaruSan** [comment]
> how we're dealing with potential mods that are uploaded shortly before deploying this code, and don't have a thumbnail generated yet

As a workaround, we could also check/generate the thumbnail when the mod page is loaded. That way the listing pages would be free of `isfile` checks, but if any mod is missing a thumbnail, then all we would have to do is load its mod page.

**Comment by DasSkelett** [comment]
Mmh, I don't really like adding a permanent workaround that adds back expensive `isfile` calls that we are working so hard to remove, for something that probably doesn't happen at all, or if, only for one or two mods. We made the mod pages storage access free in #370, would be nice if we could keep it this way.

Instead I think I am going to change the database schema after all, and add a column for the thumbnail path.
Then, in `get_or_create()`, we can first ask the db for the path, if it's not filled in yet: check if the file exists, if yes, put the path in the database and return it, if not, generate it, put the path in the db, return it.

**Comment by DasSkelett** [comment]
Okay, I think I managed to work something out, following the logic of my previous comment. It should also handle all there weird edge cases we can have, like background path in db but no file on disk.
Thumbnail is now generated on upload, but for existing ones it falls back to file system detection if they aren't in the db yet.
This should reduce file storage mount access on simple GET requests for our browse pages to a minimum.
It's also removing all the `/content/` prefixes from background paths during the db migration, so we could get rid of the ugly code that handled these cases.

Will do a squash before merge.

File: alembic/versions/2021_07_29_18_00_833ba1ee8c72.py
Comment (HebaruSan) [inline alembic/versions/2021_07_29_18_00_833ba1ee8c72.py:20]: Could we use `'` instead of `\"` as in the previous line? Slightly easier to read that way, I'd think.

Comment (DasSkelett) [reply alembic/versions/2021_07_29_18_00_833ba1ee8c72.py:20]: No, double quotes and single quotes have different meanings in SQL. Double quotes are for column and table names, single quotes for strings. Replacing them with single quotes doesn't work, and replacing the single quotes with double quotes doesn't work either, it throws syntax errors.

Comment (DasSkelett) [reply alembic/versions/2021_07_29_18_00_833ba1ee8c72.py:20]: We need the double quotes for `user`, because `user` is also an SQL instruction. And when double-quoting the table name, you apparently need to double quote all the column names as well.

Comment (DasSkelett) [reply alembic/versions/2021_07_29_18_00_833ba1ee8c72.py:20]: I guess I could try triple quotes on the outside? Not sure if they're handled specially by Python though.

Comment (HebaruSan) [reply alembic/versions/2021_07_29_18_00_833ba1ee8c72.py:20]: OK, that's fine, this isn't a big problem, just a question I had while reviewing. Never mind!

Comment (DasSkelett) [reply alembic/versions/2021_07_29_18_00_833ba1ee8c72.py]: Changed to triple quotes regardless :laughing: 
I do agree that escaped strings are hard to read, and it does appear to work just fine.

File: KerbalStuff/blueprints/profile.py
Comment (HebaruSan) [inline KerbalStuff/blueprints/profile.py:74]: `profile.thumbnail` throws an exception.

image

I assume this is left over from an earlier revision where that column would have been added.

Comment (DasSkelett) [reply KerbalStuff/blueprints/profile.py:74]: Oh yeah, that should be `profile.backgroundMedia`. Must have accidentally changed that when refactoring for the mod thumbnails. Fortunately users don't have thumbnails, that would make it all even more complicated.

I also have to take a look at that `os.path.join` again, I'm not sure if it belongs there.

File: frontend/coffee/edit_mod.coffee
Comment (HebaruSan) [inline frontend/coffee/edit_mod.coffee:26]: I think `xhr.open('POST', "/api/mod/#{window.mod_id}/update-bg")` is being evaluated by the wrong route:

image

That text is only returned by `/api/mod/<int:mod_id>/<version>`:

https://github.com/KSP-SpaceDock/SpaceDock/blob/3197cf7b3d83fd33076e8a9e63ccc6fbb28e7286/KerbalStuff/blueprints/api.py#L408

Comment (DasSkelett) [reply frontend/coffee/edit_mod.coffee:26]: Yes, you are doing a GET request there, which won't hit the correct route.
If you do a POST by actually trying to upload a background image, it calls the correct function.
However, as you said, there's indeed a bug, in that it breaks if there was no image before and `old_path` is `None`.

Fix will be pushed in a minute :+1:

**Review by HebaruSan — APPROVED** [review]
I think this is all working now.
Might as well squash and merge.

**Pull request merged**
